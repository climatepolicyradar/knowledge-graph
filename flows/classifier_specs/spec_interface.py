from pathlib import Path
from typing import Optional

import yaml
from pydantic import BaseModel, Field

from scripts.cloud import AwsEnv
from src.identifiers import ClassifierID, WikibaseID
from src.version import Version

SPEC_DIR = Path("flows") / "classifier_specs" / "v2"


class ClassifierSpec(BaseModel):
    """Details for a classifier to run."""

    class ComputeEnvironment(BaseModel):
        """Details about the compute environment the classifier should on."""

        gpu: bool = Field(
            description=("Whether the classifier should be run on a GPU."),
            default=False,
        )

    wikibase_id: WikibaseID = Field(
        description=(
            "The wikibase id for the underlying concept being classified. e.g. 'Q992'"
        ),
    )
    classifier_id: ClassifierID = Field(
        description=(
            "The unique identifier for the classifier, built from its internals."
        ),
    )
    wandb_registry_version: Version = Field(
        description=("The version of the classifier in wandb registry. e.g. v1"),
    )
    compute_environment: Optional[ComputeEnvironment] = Field(
        description=ComputeEnvironment.__doc__,
        default=None,
    )

    def __hash__(self):
        """Make ClassifierSpec hashable for use in sets and as dict keys."""
        return hash(self.classifier_id)

    def __str__(self):
        """Return a string representation of the classifier spec."""
        return f"{self.wikibase_id}:{self.classifier_id}"


def determine_spec_file_path(aws_env: AwsEnv, spec_dir: Path = SPEC_DIR) -> Path:
    """Determine the path to the spec file for a given AWS environment."""
    return spec_dir / f"{aws_env}.yaml"


def load_classifier_specs(
    aws_env: AwsEnv, spec_dir: Path = SPEC_DIR
) -> list[ClassifierSpec]:
    """Load classifier specs into python for a given environment."""
    file_path = determine_spec_file_path(aws_env, spec_dir)

    with open(file_path, "r") as file:
        contents = yaml.load(file, Loader=yaml.FullLoader)

    classifier_specs = []
    for spec in contents:
        classifier_specs.append(ClassifierSpec.model_validate(spec))

    return classifier_specs
