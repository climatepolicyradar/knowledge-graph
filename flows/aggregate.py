import json
import os
from collections.abc import AsyncGenerator, Iterable, Sequence
from typing import Any, TypeAlias, TypeVar

import aioboto3
import prefect.tasks as tasks
from botocore.exceptions import ClientError
from prefect import flow
from prefect.artifacts import create_markdown_artifact, create_table_artifact
from prefect.assets import materialize
from prefect.client.schemas.objects import FlowRun
from prefect.context import TaskRunContext, get_run_context
from prefect.futures import PrefectFuture, PrefectFutureList
from prefect.task_runners import ThreadPoolTaskRunner
from prefect.utilities.names import generate_slug
from pydantic import PositiveInt
from types_aiobotocore_s3.client import S3Client

from flows.boundary import (
    TextBlockId,
    convert_labelled_passage_to_concepts,
    s3_copy_file,
    s3_object_write_text_async,
)
from flows.classifier_specs.spec_interface import (
    ClassifierSpec,
    load_classifier_specs,
    should_skip_doc,
)
from flows.config import Config
from flows.inference import (
    deserialise_pydantic_list_with_fallback,
)
from flows.utils import (
    DocumentStem,
    S3Uri,
    SlackNotify,
    collect_unique_file_stems_under_prefix,
    iterate_batch,
    map_as_sub_flow,
)
from src.cloud import AwsEnv
from src.labelled_passage import LabelledPassage

T = TypeVar("T")
R = TypeVar("R")


DEFAULT_N_DOCUMENTS_IN_BATCH: PositiveInt = 20
DEFAULT_N_BATCHES: PositiveInt = 5

# A unique identifier for the run output made from the run context
RunOutputIdentifier: TypeAlias = str

# A string representation of a classifier spec (i.e. Q123:v4)
SpecStr: TypeAlias = str

# A serialised vespa concept, see cpr_sdk.models.search.Concept
SerialisedVespaConcept: TypeAlias = list[dict[str, str]]


class AggregationFailure(Exception):
    """A document failure."""

    def __init__(self, document_stem: DocumentStem, exception: Exception, context: str):
        super().__init__(
            f"{document_stem} | exception: {str(exception)} | context: {context}"
        )
        self.document_stem = document_stem
        self.exception = exception
        self.context = context


class AllSkippedFailure(Exception):
    """Every classifier was skipped for this document."""


def build_run_output_identifier() -> RunOutputIdentifier:
    """Builds an identifier from the start time and name of the flow run."""
    run_context = get_run_context()
    if isinstance(run_context, TaskRunContext):
        raise ValueError("expected flow run context but got task run context")

    if run_context.flow_run is None:
        raise ValueError("run context is missing flow run")

    if run_context.flow_run.start_time is None:
        raise ValueError("flow run didn't have a start time.")

    start_time = run_context.flow_run.start_time.replace(tzinfo=None).isoformat(
        timespec="minutes"
    )
    run_name = run_context.flow_run.name
    return f"{start_time}-{run_name}"


async def get_all_labelled_passages_for_one_document(
    s3: S3Client,
    document_stem: DocumentStem,
    classifier_specs: Sequence[ClassifierSpec],
    config: Config,
) -> AsyncGenerator[tuple[ClassifierSpec, list[LabelledPassage]], None]:
    """Get the labelled passages from S3."""

    for spec in classifier_specs:
        if should_skip_doc(document_stem, spec):
            continue

        s3_uri = generate_s3_uri_input(
            cache_bucket=config.cache_bucket_str,
            document_source_prefix=config.aggregate_document_source_prefix,
            classifier_spec=spec,
            document_stem=document_stem,
        )
        response = await s3.get_object(Bucket=s3_uri.bucket, Key=s3_uri.key)
        body = await response["Body"].read()
        content = body.decode("utf-8")

        labelled_passages = deserialise_pydantic_list_with_fallback(
            content, LabelledPassage
        )
        yield spec, labelled_passages


def check_all_values_are_the_same(values: list[Any]) -> bool:
    """Check if all values are the same."""
    return len(set(values)) == 1


def validate_passages_are_same_except_concepts(passages: list[LabelledPassage]) -> None:
    """Check if passages are the same (except for metadata & spans)."""
    properties = [
        "id",
        "text",
    ]
    for property in properties:
        values = [getattr(passage, property) for passage in passages]
        if not check_all_values_are_the_same(values):
            unique_values = set(values)
            raise ValueError(
                f"Found a discrepancy in passage for {property}: {unique_values}"
            )


def task_input_hash(
    remove: Sequence[str],
    context: TaskRunContext,
    arguments: dict[str, Any],
) -> str | None:
    """
    Remove arguments from a task's input.

    Sometimes we don't want to, or can't, serialise them.
    """

    return tasks.task_input_hash(
        context,
        {k: v for k, v in arguments.items() if k not in remove},
    )


def generate_s3_uri_input(
    cache_bucket: str,
    document_source_prefix: str,
    classifier_spec: ClassifierSpec,
    document_stem: DocumentStem,
) -> S3Uri:
    return S3Uri(
        bucket=cache_bucket,
        key=os.path.join(
            document_source_prefix,
            classifier_spec.wikibase_id,
            classifier_spec.classifier_id,
            f"{document_stem}.json",
        ),
    )


def generate_s3_uri_output(
    cache_bucket: str,
    aggregate_inference_results_prefix: str,
    run_output_identifier: RunOutputIdentifier,
    document_stem: DocumentStem,
) -> S3Uri:
    return S3Uri(
        bucket=cache_bucket,
        key=os.path.join(
            aggregate_inference_results_prefix,
            run_output_identifier,
            f"{document_stem}.json",
        ),
    )


def task_run_name(parameters: dict[str, Any]) -> str:
    document_id = parameters.get("document_id", "unknown")
    slug = generate_slug(2)
    return f"aggregate-single-document-{document_id}-{slug}"


@materialize(
    "foo://bar",  # Asset key is not known yet
    task_run_name=task_run_name,  # pyright: ignore[reportArgumentType]
    retries=1,
    persist_result=False,
)
async def process_document(
    document_stem: DocumentStem,
    classifier_specs: Sequence[ClassifierSpec],
    config: Config,
    run_output_identifier: RunOutputIdentifier,
) -> DocumentStem:
    """Process a single document and return its status."""
    try:
        session = aioboto3.Session(region_name=config.bucket_region)
        async with session.client("s3") as s3:
            print("Fetching labelled passages for", document_stem)

            concepts_for_vespa: dict[TextBlockId, SerialisedVespaConcept] = {}
            async for (
                spec,
                labelled_passages,
            ) in get_all_labelled_passages_for_one_document(
                s3, document_stem, classifier_specs, config
            ):
                # `concepts_for_vespa` starts empty so Validation not
                # needed initially
                if not concepts_for_vespa:
                    for passage in labelled_passages:
                        concepts_for_vespa[TextBlockId(passage.id)] = [
                            vc.model_dump(mode="json")
                            for vc in convert_labelled_passage_to_concepts(passage)
                        ]
                    continue

                if len(labelled_passages) != len(concepts_for_vespa.keys()):
                    raise ValueError(
                        f"The number of passages diverge when appending {spec}: "
                        + f"{len(labelled_passages)=} != {len(concepts_for_vespa)=}"
                    )

                for passage, text_block_id in zip(
                    labelled_passages, concepts_for_vespa.keys()
                ):
                    if passage.id != text_block_id:
                        raise ValueError(
                            f"The text_block id diverges for {spec} when "
                            + "compared with what has been collected so far:"
                            + f"{passage.id=} != {text_block_id=}"
                        )
                    serialised_concepts = [
                        vc.model_dump(mode="json")
                        for vc in convert_labelled_passage_to_concepts(passage)
                    ]
                    concepts_for_vespa[TextBlockId(passage.id)].extend(
                        serialised_concepts
                    )

            # Validation, checking if we skipped all concepts for a document
            if not concepts_for_vespa:
                raise AllSkippedFailure

            # Write to s3
            s3_uri = generate_s3_uri_output(
                cache_bucket=config.cache_bucket_str,
                aggregate_inference_results_prefix=config.aggregate_inference_results_prefix,
                run_output_identifier=run_output_identifier,
                document_stem=document_stem,
            )
            await s3_object_write_text_async(s3, s3_uri, json.dumps(concepts_for_vespa))

            # Duplicate to latest
            await s3_copy_file(
                s3,
                source=s3_uri,
                target=S3Uri(
                    bucket=config.cache_bucket_str,
                    key=os.path.join(
                        config.aggregate_inference_results_prefix,
                        "latest",
                        f"{document_stem}.json",
                    ),
                ),
            )
            return document_stem
    except AllSkippedFailure as e:
        raise AggregationFailure(
            document_stem=document_stem, exception=e, context=repr(e)
        )
    except ClientError as e:
        print(f"ClientError: {e.response}")
        raise AggregationFailure(
            document_stem=document_stem,
            exception=e,
            context=str(e.response),
        )
    except Exception as e:
        raise AggregationFailure(
            document_stem=document_stem, exception=e, context=repr(e)
        )


def handle_results(
    results: Sequence[DocumentStem | AggregationFailure],
) -> tuple[list[DocumentStem], list[AggregationFailure]]:
    success_stems: list[DocumentStem] = []
    failures: list[AggregationFailure] = []

    for result in results:
        if isinstance(result, AggregationFailure):
            failures.append(result)
        elif isinstance(result, str):
            success_stems.append(DocumentStem(result))
        else:
            raise ValueError(f"Unknown result type: {type(result)}")

    return success_stems, failures


async def create_aggregate_inference_summary_artifact(
    config: Config,
    success_stems: list[DocumentStem],
    failures: list[AggregationFailure],
) -> None:
    """Create a summary artifact of the aggregated inference results."""

    overview_description = f"""# Aggregate Inference Summary

## Overview
- **Environment**: {config.aws_env.value}
- **Documents processed**: {len(success_stems)}
- **Failed documents**: {len(failures)}/{len(success_stems)}
"""

    details = [
        {
            "Failed document Stem": failure.document_stem,
            "Exception": str(failure.exception),
            "Context": failure.context,
        }
        for failure in failures
    ]

    await create_table_artifact(  # pyright: ignore[reportGeneralTypeIssues]
        key=f"aggregate-inference-{config.aws_env.value}",
        table=details,
        description=overview_description,
    )


async def create_aggregate_inference_overall_summary_artifact(
    aws_env: AwsEnv,
    document_stems: Sequence[DocumentStem],
    classifier_specs: list[ClassifierSpec],
    run_output_identifier: RunOutputIdentifier,
    successes: Sequence[FlowRun],
    failures: Sequence[BaseException | FlowRun],
) -> None:
    """Create a summary artifact of the overall aggregated inference results."""
    markdown_content = f"""# Aggregate Inference Overall Summary

## Overview
- **Environment**: {aws_env.value}
- **Run Output Identifier**: `{run_output_identifier}`
- **Total classifier specs.**: {len(classifier_specs)}
- **Total documents**: {len(document_stems)}
- **Successful batches**: {len(successes)}
- **Failed batches**: {len(failures)}
"""

    await create_markdown_artifact(  # pyright: ignore[reportGeneralTypeIssues]
        key=f"aggregate-inference-overall-{aws_env.value}",
        markdown=markdown_content,
    )


def collect_stems_by_specs(config: Config) -> list[DocumentStem]:
    """Collect the stems for the given specs."""
    document_stems = []
    specs = load_classifier_specs(config.aws_env)
    for spec in specs:
        prefix = os.path.join(
            config.aggregate_document_source_prefix,
            spec.wikibase_id,
            spec.classifier_id,
        )
        document_stems.extend(
            collect_unique_file_stems_under_prefix(
                bucket_name=config.cache_bucket_str,
                prefix=prefix,
            )
        )

    return list(set(document_stems))


@flow(  # pyright: ignore[reportCallIssue]
    timeout_seconds=None,
    log_prints=True,
    task_runner=ThreadPoolTaskRunner(
        # This is valid, as per the docs[1].
        #
        # [1]: https://github.com/PrefectHQ/prefect/blob/01f9d5e7d5204f5b8760b431d72db52dd78e6aca/docs/v3/concepts/task-runners.mdx#L49
        max_workers=DEFAULT_N_DOCUMENTS_IN_BATCH  # pyright: ignore[reportArgumentType]
    ),
)
async def aggregate_batch_of_documents(
    document_stems: Sequence[DocumentStem],
    config_json: dict[str, Any],
    classifier_specs: Sequence[ClassifierSpec],
    run_output_identifier: RunOutputIdentifier,
) -> RunOutputIdentifier:
    """Aggregate the inference results for the given document ids."""
    config = Config.model_validate(config_json)

    tasks: list[PrefectFuture[DocumentStem]] = []

    print("submitting tasks")
    for document_stem in document_stems:
        assets: Sequence[str] = [
            str(
                generate_s3_uri_output(
                    cache_bucket=config.cache_bucket_str,
                    aggregate_inference_results_prefix=config.aggregate_inference_results_prefix,
                    run_output_identifier=run_output_identifier,
                    document_stem=document_stem,
                )
            )
        ]
        asset_deps: list[str] = [
            str(
                generate_s3_uri_input(
                    cache_bucket=config.cache_bucket_str,
                    document_source_prefix=config.aggregate_document_source_prefix,
                    classifier_spec=spec,
                    document_stem=document_stem,
                )
            )
            for spec in classifier_specs
        ]

        tasks.append(
            process_document.with_options(  # pyright: ignore[reportFunctionMemberAccess, reportArgumentType]
                assets=assets,
                asset_deps=asset_deps,  # pyright: ignore[reportArgumentType]
            ).submit(
                document_stem=document_stem,
                classifier_specs=classifier_specs,
                config=config,
                run_output_identifier=run_output_identifier,
            )
        )

    # This is valid, as per the source code[1].
    #
    # [1]: https://github.com/PrefectHQ/prefect/blob/01f9d5e7d5204f5b8760b431d72db52dd78e6aca/src/prefect/task_runners.py#L183-L213
    futures: PrefectFutureList[DocumentStem] = PrefectFutureList(tasks)  # pyright: ignore[reportAbstractUsage]

    print("getting results")
    results = futures.result(raise_on_failure=False)

    print("handling results")
    success_stems, failures = handle_results(results)

    print("creating summary artifact")
    await create_aggregate_inference_summary_artifact(
        config=config,
        success_stems=success_stems,
        failures=failures,
    )

    if failures:
        raise ValueError(
            f"Saw {len(failures)} failures when aggregating inference results"
        )

    return run_output_identifier


@flow(
    on_failure=[SlackNotify.message],
    on_crashed=[SlackNotify.message],
    timeout_seconds=None,
    log_prints=True,
)
async def aggregate(
    document_stems: None | Sequence[DocumentStem] = None,
    config: Config | None = None,
    n_documents_in_batch: PositiveInt = DEFAULT_N_DOCUMENTS_IN_BATCH,
    n_batches: PositiveInt = DEFAULT_N_BATCHES,
) -> RunOutputIdentifier:
    """Aggregate the inference results for the given document ids."""
    if not config:
        print("no config provided, creating one")
        config = await Config.create()

    if not document_stems:
        print(
            "no document stems provided, collecting all available from S3 under prefix: "
            + f"{config.aggregate_document_source_prefix}"
        )
        document_stems = collect_stems_by_specs(config)

    run_output_identifier = build_run_output_identifier()
    classifier_specs = load_classifier_specs(config.aws_env)

    print(
        f"Aggregating inference results for {len(document_stems)} documents, using "
        + f"{len(classifier_specs)} classifiers, outputting to {run_output_identifier}"
    )

    batches = iterate_batch(
        document_stems,
        n_documents_in_batch,
    )

    def parameters(batch: Sequence[DocumentStem]) -> dict[str, Any]:
        return {
            "document_stems": batch,
            "config_json": config.model_dump(),
            "classifier_specs": classifier_specs,
            "run_output_identifier": run_output_identifier,
        }

    parameterised_batches: Iterable[dict[str, Any]] = (
        parameters(batch) for batch in batches
    )

    successes, failures = await map_as_sub_flow(  # pyright: ignore[reportCallIssue]
        # The typing doesn't pick up the Flow decorator
        fn=aggregate_batch_of_documents,  # pyright: ignore[reportArgumentType]
        aws_env=config.aws_env,
        counter=n_batches,
        parameterised_batches=parameterised_batches,
        unwrap_result=False,
    )

    await create_aggregate_inference_overall_summary_artifact(
        aws_env=config.aws_env,
        document_stems=document_stems,
        classifier_specs=classifier_specs,
        run_output_identifier=run_output_identifier,
        successes=successes,
        failures=failures,
    )

    if failures:
        raise ValueError(f"{len(failures)}/{len(failures) + len(successes)} failed")

    return run_output_identifier
