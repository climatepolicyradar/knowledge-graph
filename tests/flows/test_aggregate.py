import asyncio
import json
import os
import tempfile
import uuid
from pathlib import Path
from unittest.mock import MagicMock, patch

import pydantic
import pytest
from cpr_sdk.models.search import Concept as VespaConcept
from prefect import flow
from prefect.artifacts import Artifact
from prefect.client.schemas.objects import FlowRun
from prefect.context import FlowRunContext
from prefect.states import Running

from flows.aggregate import (
    AggregationFailure,
    Metadata,
    aggregate_batch_of_documents,
    build_run_output_identifier,
    collect_stems_by_specs,
    get_all_labelled_passages_for_one_document,
    process_document,
    store_metadata,
    validate_passages_are_same_except_concepts,
)
from flows.classifier_specs.spec_interface import ClassifierSpec
from flows.utils import DocumentStem
from knowledge_graph.identifiers import ConceptID, WikibaseID
from knowledge_graph.labelled_passage import LabelledPassage
from knowledge_graph.span import Span
from scripts.update_classifier_spec import write_spec_file


@pytest.fixture
def mock_classifier_specs():
    with tempfile.TemporaryDirectory() as spec_dir:
        # Write the concept specs to a YAML file
        temp_spec_dir = Path(spec_dir)
        classifier_specs = [
            ClassifierSpec(
                wikibase_id="Q123",
                classifier_id="g29kcna9",
                wandb_registry_version="v4",
                dont_run_on=["sabin"],
            ),
            ClassifierSpec(
                wikibase_id="Q218",
                classifier_id="6z4pufsm",
                wandb_registry_version="v5",
                dont_run_on=["sabin"],
            ),
            ClassifierSpec(
                wikibase_id="Q223",
                classifier_id="36bhx4mu",
                wandb_registry_version="v3",
                dont_run_on=["sabin"],
            ),
            ClassifierSpec(
                wikibase_id="Q767",
                classifier_id="mgwutbqx",
                wandb_registry_version="v3",
                dont_run_on=["sabin"],
            ),
            ClassifierSpec(
                wikibase_id="Q1286",
                classifier_id="7bt99yeu",
                wandb_registry_version="v3",
                dont_run_on=["sabin", "cclw"],
            ),
        ]
        spec_file_path = temp_spec_dir / "sandbox.yaml"
        write_spec_file(spec_file_path, classifier_specs)
        with patch("flows.classifier_specs.spec_interface.SPEC_DIR", temp_spec_dir):
            yield spec_file_path, classifier_specs


@pytest.mark.asyncio
async def test_aggregate_batch_of_documents(
    mock_bucket_labelled_passages_large, test_config, mock_classifier_specs
):
    _, bucket, s3_async_client = mock_bucket_labelled_passages_large
    classifier_specs: list[ClassifierSpec] = mock_classifier_specs[1]

    document_stems = [
        DocumentStem("CCLW.executive.10061.4515"),
        DocumentStem("CPR.document.i00000549.n0000"),
        DocumentStem("UNFCCC.non-party.467.0"),
        DocumentStem("UNFCCC.party.492.0"),
    ]

    run_reference = await aggregate_batch_of_documents(
        document_stems=document_stems,
        config_json=test_config.model_dump(),
        classifier_specs=classifier_specs,
        run_output_identifier="test-run",
    )

    all_collected_ids = []
    for document_stem in document_stems:
        s3_path = os.path.join(
            test_config.aggregate_inference_results_prefix,
            run_reference,
            f"{document_stem}.json",
        )
        try:
            response = await s3_async_client.get_object(Bucket=bucket, Key=s3_path)
            data = await response["Body"].read()
            data = json.loads(data.decode("utf-8"))
        except s3_async_client.exceptions.NoSuchKey:
            pytest.fail(f"Unable to find output file: {s3_path}")
        except json.JSONDecodeError:
            pytest.fail(f"Unable to deserialise output for: {document_stem}")
        except Exception as e:
            pytest.fail(f"Unexpected error: {e}")

        wikibase_ids = [
            classifier_spec.wikibase_id for classifier_spec in classifier_specs
        ]

        document_inference_output = list(data.values())
        collected_ids_for_document = []
        for concepts in document_inference_output:
            for concept in concepts:
                try:
                    vespa_concept = VespaConcept.model_validate(concept)
                    collected_ids_for_document.append(vespa_concept.id)
                except pydantic.ValidationError as e:
                    pytest.fail(
                        f"Unable to deserialise concept: {concept} with error: {e}"
                    )

        # Q1286 should not be on CCLW.executive.10061.4515 because of dont_run_on
        if "CCLW" in document_stem:
            assert "Q1286" not in collected_ids_for_document
        assert len(collected_ids_for_document) > 0, (
            f"No concepts found for document: {document_stem}"
        )
        all_collected_ids.extend(collected_ids_for_document)

    assert set(all_collected_ids) == set(wikibase_ids), (
        f"Outputted: {set(all_collected_ids)} which doesnt match those in the specs: {set(wikibase_ids)}"
    )
    COUNT = 139
    assert len(all_collected_ids) == COUNT, (
        f"Expected {COUNT} concepts to be outputted, found: {len(all_collected_ids)}"
    )

    summary_artifact = await Artifact.get("aggregate-inference-sandbox")
    assert summary_artifact and summary_artifact.description
    assert summary_artifact.data == "[]"


@pytest.mark.asyncio
async def test_aggregate_batch_of_documents__with_failures(
    mock_bucket_labelled_passages_large, mock_classifier_specs, test_config
):
    _, classifier_specs = mock_classifier_specs
    expect_failure_stems = [
        DocumentStem("CCLW.Made.Up.Document.ID"),
        DocumentStem("OEP.One.That.Should.Fail"),
    ]
    document_stems = [DocumentStem("CCLW.executive.10061.4515")] + expect_failure_stems

    with pytest.raises(ValueError):
        await aggregate_batch_of_documents(
            document_stems=document_stems,
            config_json=test_config.model_dump(),
            classifier_specs=classifier_specs,
            run_output_identifier="test-run",
        )

    summary_artifact = await Artifact.get("aggregate-inference-sandbox")
    assert summary_artifact and summary_artifact.description
    artifact_data = json.loads(summary_artifact.data)
    failure_stems = [f["Failed document Stem"] for f in artifact_data]
    assert set(failure_stems) == set(expect_failure_stems)
    assert "NoSuchKey" in artifact_data[0]["Exception"]
    assert "NoSuchKey" in artifact_data[1]["Exception"]


def test_build_run_output_prefix():
    @flow()
    def fake_flow():
        return build_run_output_identifier()

    prefix = fake_flow()
    # From https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
    s3_unsupported_chars = r'\^`>{[<%#"}]~|'

    for char in s3_unsupported_chars:
        assert char not in prefix, f"Unsupported char found in prefix: {prefix}"

    assert "None" not in prefix


@pytest.mark.asyncio
async def test_get_all_labelled_passages_for_one_document(
    mock_bucket_labelled_passages_large, test_config
):
    _, _, s3_async_client = mock_bucket_labelled_passages_large
    document_stem = DocumentStem("CCLW.executive.10061.4515")
    classifier_specs = [
        ClassifierSpec(
            wikibase_id="Q218", classifier_id="6z4pufsm", wandb_registry_version="v5"
        ),
        ClassifierSpec(
            wikibase_id="Q767", classifier_id="mgwutbqx", wandb_registry_version="v3"
        ),
        ClassifierSpec(
            wikibase_id="Q1286", classifier_id="7bt99yeu", wandb_registry_version="v3"
        ),
        ClassifierSpec(
            wikibase_id="Q1400",
            classifier_id="7bt99yeu",
            wandb_registry_version="v3",
            dont_run_on=["cclw"],
        ),
    ]
    all_labelled_passages = []
    async for spec, labelled_passages in get_all_labelled_passages_for_one_document(
        s3_async_client, document_stem, classifier_specs, test_config
    ):
        all_labelled_passages.append(labelled_passages)
    assert len(all_labelled_passages) == 3


def test_validate_passages_are_same_except_concepts():
    passages: list[LabelledPassage] = []
    for i in range(10):
        span_one = Span(text=f"unique spans: {i}", start_index=0, end_index=4)
        span_two = Span(text=f"unique spans: {i} {i}", start_index=0, end_index=4)
        passage = LabelledPassage(
            id="1",
            text="id and text should match across identical labelled passages",
            spans=[span_one, span_two],
        )
        passages.append(passage)

    validate_passages_are_same_except_concepts(passages)

    with pytest.raises(ValueError):
        passages.append(
            LabelledPassage(
                id="2",
                text="Imagine if we messed up and drew in a different passage",
                spans=[Span(text="span", start_index=0, end_index=4)],
            )
        )
        validate_passages_are_same_except_concepts(passages)


@pytest.mark.asyncio
async def test_process_single_document__success(
    mock_bucket_labelled_passages_large, mock_classifier_specs, test_config
):
    _, bucket, s3_async_client = mock_bucket_labelled_passages_large
    _, classifier_specs = mock_classifier_specs
    document_stem = DocumentStem("CCLW.executive.10061.4515")

    assert document_stem == await process_document.fn(
        document_stem,
        classifier_specs,
        test_config,
        "run_output_identifier",
    )

    # Check that the file is in the run_output_identifier prefix with head object
    response = await s3_async_client.head_object(
        Bucket=bucket,
        Key=os.path.join(
            test_config.aggregate_inference_results_prefix,
            "run_output_identifier",
            f"{document_stem}.json",
        ),
    )
    assert response["ContentLength"] > 0

    # Check that the file is in the latest prefix with head object
    response = await s3_async_client.head_object(
        Bucket=bucket,
        Key=os.path.join(
            test_config.aggregate_inference_results_prefix,
            "latest",
            f"{document_stem}.json",
        ),
    )
    assert response["ContentLength"] > 0


@pytest.mark.asyncio
async def test_process_single_document__client_error(
    mock_bucket_labelled_passages_large, mock_classifier_specs, test_config
):
    document_stem = DocumentStem("CCLW.executive.10061.4515")
    spec_file_path, classifier_specs = mock_classifier_specs
    classifier_specs.append(
        ClassifierSpec(
            wikibase_id="Q9999999999",
            classifier_id="zzzzzzzz",
            wandb_registry_version="v99",
        )
    )
    write_spec_file(spec_file_path, classifier_specs)

    result = await asyncio.gather(
        process_document.fn(
            document_stem,
            classifier_specs,
            test_config,
            "run_output_identifier",
        ),
        return_exceptions=True,
    )
    assert len(result) == 1
    assert isinstance(result[0], AggregationFailure)
    code = result[0].exception.response["Error"]["Code"]
    assert code == "NoSuchKey"
    key = result[0].exception.response["Error"]["Key"]
    assert (
        key == "labelled_passages/Q9999999999/zzzzzzzz/CCLW.executive.10061.4515.json"
    )


@pytest.mark.asyncio
async def test_process_single_document__value_error(
    mock_bucket_labelled_passages_large,
    mock_classifier_specs,
    test_config,
    snapshot,
):
    keys, bucket, s3_async_client = mock_bucket_labelled_passages_large
    _, classifier_specs = mock_classifier_specs

    # Use Q223 and Q767 to create mismatch.
    #
    # Those classifiers are setup in the fixtures.
    classifier_specs = [
        spec for spec in classifier_specs if spec.wikibase_id in ["Q223", "Q767"]
    ]

    document_stem = DocumentStem("CCLW.executive.10061.4515")

    # Replace Q767 with fewer passages to create mismatch with Q223
    # (which has 27).
    new_data_short = [
        '{"id":"b1","text":"Some words","spans":[],"metadata":{}}',
        '{"id":"b2","text":"But not enough words","spans":[],"metadata":{}}',
    ]

    # Find and replace the Q767 file to have only 2 passages, while
    # Q223 keeps its 27.
    q767_key = [
        k for k in keys if "CCLW.executive.10061.4515" in k and "Q767/mgwutbqx" in k
    ][0]
    await s3_async_client.put_object(
        Bucket=bucket,
        Key=q767_key,
        Body=json.dumps(new_data_short),
    )

    async with asyncio.timeout(5):
        result = await asyncio.gather(
            process_document.fn(
                document_stem,
                classifier_specs,
                test_config,
                "run_output_identifier",
            ),
            return_exceptions=True,
        )
    assert result == snapshot


@pytest.mark.asyncio
async def test_collect_stems_by_specs(
    mock_classifier_specs, mock_bucket_labelled_passages_large, test_config
):
    stems = await collect_stems_by_specs(test_config)
    assert set(stems) == set(
        [
            "UNFCCC.non-party.467.0",
            "CCLW.executive.10061.4515",
            "AF.document.i00000021.n0000_translated_en",
            "UNFCCC.party.492.0",
            "CPR.document.i00000549.n0000",
        ]
    )


@pytest.mark.asyncio
async def test_store_metadata(
    test_config,
    mock_async_bucket,
    mock_s3_async_client,
    snapshot,
):
    """Test that store_metadata correctly builds S3 URI and stores metadata."""
    mock_tags = ["tag:value1", "sha:abc123", "branch:main"]
    mock_run_output_id = "2025-01-15T10:30-test-flow-run"

    # Create a real FlowRun object with proper data
    flow_run = FlowRun(
        id=uuid.UUID("0199bef8-7e41-7afc-9b4c-d3abd406be84"),
        flow_id=uuid.UUID("b213352f-3214-48e3-8f5d-ec19959cb28e"),
        name="test-flow-run",
        state=Running(),
        tags=mock_tags,
    )

    mock_context = MagicMock(spec=FlowRunContext)
    mock_context.flow_run = flow_run

    classifier_specs = [
        ClassifierSpec(
            concept_id=ConceptID("xyz78abc"),
            wikibase_id=WikibaseID("Q788"),
            classifier_id="abcd2345",
            wandb_registry_version="v1",
        )
    ]

    # Mock only the Prefect context, let moto handle S3
    with patch("flows.aggregate.get_run_context", return_value=mock_context):
        await store_metadata(
            config=test_config,
            classifier_specs=classifier_specs,
            run_output_identifier=mock_run_output_id,
        )

    expected_key = os.path.join(
        test_config.aggregate_inference_results_prefix,
        mock_run_output_id,
        "metadata.json",
    )

    response = await mock_s3_async_client.head_object(
        Bucket=test_config.cache_bucket, Key=expected_key
    )
    assert response["ContentLength"] > 0, (
        f"Expected S3 object {expected_key} to have content"
    )

    response = await mock_s3_async_client.get_object(
        Bucket=test_config.cache_bucket, Key=expected_key
    )
    metadata_content = await response["Body"].read()
    metadata_dict = json.loads(metadata_content.decode("utf-8"))

    metadata = Metadata.model_validate(metadata_dict)
    assert metadata == snapshot
