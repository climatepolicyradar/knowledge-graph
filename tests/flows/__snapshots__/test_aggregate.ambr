# serializer version: 1
# name: test_process_single_document__value_error
  list([
    AggregationFailure('CCLW.executive.10061.4515 | exception: The number of passages diverge when appending Q767:mgwutbqx: len(labelled_passages)=2 != len(concepts_for_vespa)=27 | context: '),
  ])
# ---
